peps/pep-0776.rst

```diff
@@ -0,0 +1,794 @@
+have the synchronous loading limit. We will continue with this approach in
+Emscripten Python.
+
+In the long run, we hope to implement a stack switching ``dlopen``, but that is
+out of scope for this PEP.
+
+Missing RPATH Support
+~~~~~~~~~~~~~~~~~~~~~
+
+Another important limitation of the Emscripten dynamic loader is that it does
+not currently have RPATH support. Pyodide's present workaround is as follows:
+``auditwheel-emscripten`` places shared library dependencies that are vendored
+into a package in a ``${package}.libs`` folder, following auditwheel's
+convention. Pyodide patches the dynamic loader to treat this ``${package}.libs``
+folder as if it were on the RPATH of all of the dynamic libraries in the wheel.
+
+In Emscripten 4.0.5, we have updated the shared object file format, ``wasm-ld``
+and ``emcc`` to accept an ``-rpath`` flag. We are still working on updating the
+dynamic loader to respect the rpath, but we expect this will be finished in the
+next Emscripten release. Pyodide will then switch to using the RPATH and drop
+the patch on the dynamic loader.
+
+Emscripten Python currently uses the unpatched dynamic loader and so cannot load
+extension modules that depend on vendored dynamic libraries via DT_NEEDED.
+Extension modules can load dynamic libraries via DT_NEEDED if they are in the
+system ``lib`` directory. We will wait to resolve this until we have fixed the
+Emscripten dynamic loader upstream. When Emscripten Python is built with a
+compatible version of Emscripten, it will automatically pick up support for
+wheels with vendored dynamic libraries.
+
+
+Traps and Uncaught Exceptions
+-----------------------------
+
+We consider the C runtime state to be corrupted if there is a WebAssembly trap,
+an unhandled JavaScript exception, or an uncaught WebAssembly throw instruction.
+
+Unlike in other platforms, there is no operating system to shut down the
+executable when there is a trap or other unrecoverable corruption of the libc
+runtime. We need to provide our own code to print tracebacks, dump the memory,
+or do whatever else is helpful for debugging a crash. If we expose a JavaScript
+API, we also must ensure that it is disabled after an unrecoverable crash to
+prevent downstream users from observing the Python runtime in an inconsistent
+state.
+
+In order to detect fatal errors, Pyodide uses the following approach: all
+fallable calls from WebAssembly into JavaScript are wrapped with a JavaScript
+try/catch block. Any caught JavaScript exceptions are translated into Python
+exceptions. This ensures that any recoverable JavaScript error is caught before
+it unwinds through any WebAssembly frames. All entrypoints to WebAssembly are
+also wrapped with JavaScript try/catch blocks. Any exceptions caught there have
+unwound WebAssembly frames and are thus considered to be fatal errors (though
+there is a special case to handle ``exit()``). This requires foundational
```

Comment:
```suggestion
there is a special case to handle :func:`~sys.exit()`). This requires foundational
```

Reply:
I was really thinking of the libc exit function here. But certainly the most likely the reason it would be called is because someone called `sys.exit()`.